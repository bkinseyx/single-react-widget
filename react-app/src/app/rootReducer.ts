import { combineReducers } from '@reduxjs/toolkit';

import counterReducer from '../features/reduxDemo/reduxDemoSlice';
import profileReducer from '../features/profileCard/profileSlice';
import toDoReducer from '../features/toDoCard/toDoSlice';
import userReducer from '../features/userCard/userSlice';

export const rootReducer = combineReducers({
  reduxDemo: counterReducer,
  profile: profileReducer,
  toDo: toDoReducer,
  user: userReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
