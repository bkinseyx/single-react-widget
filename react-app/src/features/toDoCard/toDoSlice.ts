import { createSlice } from '@reduxjs/toolkit';

import { createToDoReducer } from './reducers/createToDoReducer';
import { deleteToDoReducer } from './reducers/deleteToDoReducer';
import { updateToDoReducer } from './reducers/updateToDoReducer';
import { resetReducer } from './reducers/resetReducer';
import { setErrorMessageReducer } from './reducers/setErrorMessageReducer';

export interface ToDo {
  toDoId: number;
  description?: string;
  profileId: number;
  isComplete: boolean;
}

export interface ToDoState {
  toDoList: ToDo[];
  maxToDoId: number;
  errorMessage?: string;
}

export const initialState: ToDoState = {
  toDoList: [
    { toDoId: 1, profileId: 1, description: 'eat tacos', isComplete: true },
    { toDoId: 2, profileId: 1, description: 'drink milk', isComplete: false },
    {
      toDoId: 3,
      profileId: 1,
      description: 'walk and chew gum',
      isComplete: false,
    },
  ],
  maxToDoId: 3,
};

export const toDoSlice = createSlice({
  name: 'toDo',
  initialState,
  reducers: {
    createToDo: createToDoReducer,
    deleteToDo: deleteToDoReducer,
    updateToDo: updateToDoReducer,
    setErrorMessage: setErrorMessageReducer,
    reset: resetReducer,
  },
});

export const {
  createToDo,
  deleteToDo,
  updateToDo,
  setErrorMessage,
  reset,
} = toDoSlice.actions;

export default toDoSlice.reducer;
