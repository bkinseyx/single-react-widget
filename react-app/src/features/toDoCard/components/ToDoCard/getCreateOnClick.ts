import { store } from 'app/store';
import { createToDo, setErrorMessage } from '../../toDoSlice';

export const getCreateOnClick = (profileId?: number) => (): void => {
  if (!profileId) {
    store.dispatch(setErrorMessage('Must select a profile first.'));
    return;
  }
  store.dispatch(createToDo(profileId));
};
