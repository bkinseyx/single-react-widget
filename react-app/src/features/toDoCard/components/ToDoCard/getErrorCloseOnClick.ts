import { store } from 'app/store';
import { setErrorMessage } from '../../toDoSlice';

export const getErrorCloseOnClick = () => (): void => {
  store.dispatch(setErrorMessage());
};
