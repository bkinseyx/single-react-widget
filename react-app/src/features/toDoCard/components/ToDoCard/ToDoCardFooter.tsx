import React from 'react';
import { useSelector } from 'react-redux';
import { Alert } from 'react-bootstrap';
import classNames from 'classnames';

import { CardFooterWrapper } from 'app/components/CardFooterWrapper/CardFooterWrapper';
import { getCreateOnClick } from './getCreateOnClick';
import { getErrorCloseOnClick } from './getErrorCloseOnClick';
import { activeProfileSelector } from '../../selectors/activeProfileSelector';
import { errorMessageSelector } from 'features/toDoCard/selectors/errorMessageSelector';

export const ToDoCardFooter: React.FC = () => {
  const activeProfile = useSelector(activeProfileSelector);
  const errorMessage = useSelector(errorMessageSelector);
  return (
    <div className="card-footer">
      <CardFooterWrapper>
        <button
          className={classNames('btn', 'btn-primary', {
            disabled: !activeProfile?.profileId,
          })}
          onClick={getCreateOnClick(activeProfile?.profileId)}
        >
          Create New ToDo
        </button>
        {errorMessage && (
          <Alert
            style={{ flexGrow: 1 }}
            variant="warning"
            dismissible
            onClose={getErrorCloseOnClick()}
          >
            {errorMessage}
          </Alert>
        )}
      </CardFooterWrapper>
    </div>
  );
};
