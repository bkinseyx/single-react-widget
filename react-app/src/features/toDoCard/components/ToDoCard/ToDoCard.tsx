import React from 'react';
import { useSelector } from 'react-redux';

import { ToDoItem } from '../ToDoItem/ToDoItem';
import { toDoListSelector } from '../../selectors/toDoListSelector';
import { activeProfileSelector } from '../../selectors/activeProfileSelector';
import { ToDoCardFooter } from './ToDoCardFooter';
import { TopLevelComponent } from 'app/utils/typeUtils';

interface ToDoCardProps {
  reactConfig?: {
    title?: string;
  };
}

export const ToDoCard: TopLevelComponent<ToDoCardProps> = ({ reactConfig }) => {
  const toDoList = useSelector(toDoListSelector);
  const activeProfile = useSelector(activeProfileSelector);
  return (
    <div className="card" role="complementary" aria-label="ToDo Card">
      <div className="card-header">
        <h5 className="card-title">
          {reactConfig?.title}{' '}
          {activeProfile?.name && `for ${activeProfile?.name}`}
        </h5>
      </div>
      <div className="card-body">
        <ul className="list-group">
          {toDoList.map((toDo) => (
            <ToDoItem key={toDo.toDoId} toDo={toDo}></ToDoItem>
          ))}
        </ul>
      </div>
      <ToDoCardFooter></ToDoCardFooter>
    </div>
  );
};
export default ToDoCard;
