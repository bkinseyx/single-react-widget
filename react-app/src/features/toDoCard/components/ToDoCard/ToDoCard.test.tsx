import React from 'react';
import {
  render,
  screen,
  within,
  RenderResult,
  fireEvent,
} from '@testing-library/react';
import { Provider } from 'react-redux';

import { store } from 'app/store';
import { ToDoCard } from './ToDoCard';
import { reset } from '../../toDoSlice';
import { setActiveProfile } from 'features/profileCard/profileSlice';

const renderCard = (): RenderResult =>
  render(
    <Provider store={store}>
      <ToDoCard />
    </Provider>,
  );

test('renders the ToDoCard', () => {
  const { asFragment } = renderCard();
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByLabelText(/toDo card/i)).toBeInTheDocument();
  const toDoList = screen.getByRole('list');
  expect(toDoList).toBeInTheDocument();

  // we have to use queryAll* instead of getAllBy* in order to test non-existence.
  // see https://stackoverflow.com/questions/52783144/how-do-you-test-for-the-non-existence-of-an-element-using-jest-and-react-testing
  const toDoListItems = within(toDoList).queryAllByRole('listitem');
  expect(toDoListItems).toHaveLength(0);
});

test('test delete button', () => {
  store.dispatch(setActiveProfile(1));
  const { rerender, asFragment } = renderCard();
  const toDoList = screen.getByRole('list');
  let toDoListItems = within(toDoList).getAllByRole('listitem');
  const deleteButton = within(toDoListItems[0]).getByText(/delete/i);
  fireEvent.click(deleteButton);
  rerender(
    <Provider store={store}>
      <ToDoCard />
    </Provider>,
  );
  toDoListItems = within(toDoList).getAllByRole('listitem');
  expect(toDoListItems).toHaveLength(2);
  expect(asFragment()).toMatchSnapshot();
});

test('test create button', () => {
  store.dispatch(reset());
  store.dispatch(setActiveProfile(1));
  const { rerender, asFragment } = renderCard();
  const createButton = screen.getByText(/create/i);
  const toDoList = screen.getByRole('list');
  let toDoListItems = within(toDoList).getAllByRole('listitem');
  expect(toDoListItems).toHaveLength(3);
  fireEvent.click(createButton);
  rerender(
    <Provider store={store}>
      <ToDoCard />
    </Provider>,
  );
  toDoListItems = within(toDoList).getAllByRole('listitem');
  expect(toDoListItems).toHaveLength(4);
  expect(asFragment()).toMatchSnapshot();
  const lastItem = toDoListItems[3];
  expect(within(lastItem).getByLabelText(/complete/i)).not.toBeChecked();
  expect(within(lastItem).getByLabelText(/description/i)).toHaveValue('');
});
