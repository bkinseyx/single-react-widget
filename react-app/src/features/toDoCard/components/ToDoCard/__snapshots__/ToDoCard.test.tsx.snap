// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the ToDoCard 1`] = `
<DocumentFragment>
  <div
    aria-label="ToDo Card"
    class="card"
    role="complementary"
  >
    <div
      class="card-header"
    >
      <h5
        class="card-title"
      >
        ToDos 
      </h5>
    </div>
    <div
      class="card-body"
    >
      <ul
        class="list-group"
      />
    </div>
    <div
      class="card-footer"
    >
      <div
        class="wrapper"
      >
        <button
          class="btn btn-primary disabled"
        >
          Create New ToDo
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test create button 1`] = `
<DocumentFragment>
  <div
    aria-label="ToDo Card"
    class="card"
    role="complementary"
  >
    <div
      class="card-header"
    >
      <h5
        class="card-title"
      >
        ToDos for Ben
      </h5>
    </div>
    <div
      class="card-body"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item toDoItem"
        >
          <div>
            <label
              for="isComplete-1-input"
            >
              Complete?
            </label>
            <input
              checked=""
              class="form-control checkbox"
              id="isComplete-1-input"
              type="checkbox"
            />
          </div>
          <div>
            <label
              for="description-1-input"
            >
              Description
            </label>
            <input
              class="form-control"
              id="description-1-input"
              type="text"
              value="eat tacos"
            />
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
        <li
          class="list-group-item toDoItem"
        >
          <div>
            <label
              for="isComplete-2-input"
            >
              Complete?
            </label>
            <input
              class="form-control checkbox"
              id="isComplete-2-input"
              type="checkbox"
            />
          </div>
          <div>
            <label
              for="description-2-input"
            >
              Description
            </label>
            <input
              class="form-control"
              id="description-2-input"
              type="text"
              value="drink milk"
            />
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
        <li
          class="list-group-item toDoItem"
        >
          <div>
            <label
              for="isComplete-3-input"
            >
              Complete?
            </label>
            <input
              class="form-control checkbox"
              id="isComplete-3-input"
              type="checkbox"
            />
          </div>
          <div>
            <label
              for="description-3-input"
            >
              Description
            </label>
            <input
              class="form-control"
              id="description-3-input"
              type="text"
              value="walk and chew gum"
            />
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
        <li
          class="list-group-item toDoItem"
        >
          <div>
            <label
              for="isComplete-4-input"
            >
              Complete?
            </label>
            <input
              class="form-control checkbox"
              id="isComplete-4-input"
              type="checkbox"
            />
          </div>
          <div>
            <label
              for="description-4-input"
            >
              Description
            </label>
            <input
              class="form-control"
              id="description-4-input"
              type="text"
              value=""
            />
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="card-footer"
    >
      <div
        class="wrapper"
      >
        <button
          class="btn btn-primary"
        >
          Create New ToDo
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test delete button 1`] = `
<DocumentFragment>
  <div
    aria-label="ToDo Card"
    class="card"
    role="complementary"
  >
    <div
      class="card-header"
    >
      <h5
        class="card-title"
      >
        ToDos for Ben
      </h5>
    </div>
    <div
      class="card-body"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item toDoItem"
        >
          <div>
            <label
              for="isComplete-2-input"
            >
              Complete?
            </label>
            <input
              class="form-control checkbox"
              id="isComplete-2-input"
              type="checkbox"
            />
          </div>
          <div>
            <label
              for="description-2-input"
            >
              Description
            </label>
            <input
              class="form-control"
              id="description-2-input"
              type="text"
              value="drink milk"
            />
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
        <li
          class="list-group-item toDoItem"
        >
          <div>
            <label
              for="isComplete-3-input"
            >
              Complete?
            </label>
            <input
              class="form-control checkbox"
              id="isComplete-3-input"
              type="checkbox"
            />
          </div>
          <div>
            <label
              for="description-3-input"
            >
              Description
            </label>
            <input
              class="form-control"
              id="description-3-input"
              type="text"
              value="walk and chew gum"
            />
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="card-footer"
    >
      <div
        class="wrapper"
      >
        <button
          class="btn btn-primary"
        >
          Create New ToDo
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
